client/allocrunner/alloc_runner.go

```diff
@@ -1348,6 +1348,9 @@ func (ar *allocRunner) Reconnect(update *structs.Allocation) (err error) {
 	// Build the client allocation
 	alloc := ar.clientAlloc(states)
 
+	// Set the reconnect modify index so that the scheduler can track that the reconnect has not been processed.
+	alloc.ReconnectModifyIndex = ar.Alloc().AllocModifyIndex
```

Comment:
So in client.go line 2033 it get's added to the stripped alloc during allocSync, that then gets sent to Node Update, which then updates state, and triggers an eval, When the  eval fires, the index is set. We then have to unset it when applying the plan.  Have you tried it out?  I had logging in here during development showing it all flowed through.

Reply:
> which then updates state

That's the bit where I don't see how it's happening. Any update of an existing object takes a copy first (ref [`state_store.go#L3474`](https://github.com/hashicorp/nomad/blob/main/nomad/state/state_store.go#L3474)) and then modifies the copy before inserting it. So if we haven't pulled in information that the client is authoritative on, the state isn't getting updated for the transaction. 

I haven't had a chance to test it out thoroughly (still trying to get 1.4.2 out! 😁 ) but I suspect the reason it's "working" right now is because of the state store corruption on line 1223. That'll appear correct under some circumstances but won't have gone thru raft correctly.